### DISCUSSION TITLE
v0.9.0-rc1

### DISCUSSION THREAD

**Discussion by ISSOtm** [gbdev/rgbds#1507] [Discussions]
> [!TIP]
> Starting with this release, we publish universal ("fat") binaries for macOS, so they'll support x86-64 *and* newer ARM (“Apple Silicon”) CPUs!

## :warning: Deprecated

- Treating multi-character strings as numbers (you can use a single 32-bit character mapping instead)

## Added

- RGBASM `-s` outputs a "state file", listing the values of (non-label) symbols when finished assembling
- RGBASM `-Wpurge` warns when purging exported symbols or labels
- Symbols can have non-leading `$` characters (so an interpolated `prefix{numeric}` symbol works)
- Symbols can have the same names as reserved keywords by prefixing them with a `#`
- `CHARMAP` can map characters to sequences of multiple values, with values not limited to 8 bits
- A string constant `.` (period) which expands to the current global label scope
- A string constant `..` (two periods) which expands to the current fully-qualified local label scope (e.g. `Global.local`)
- `BITWIDTH(n)` returns the number of bits necessary to represent *n*
- `TZCOUNT(n)` returns the count of trailing zero bits of *n*
- `EXPORT DEF` defines and exports a numeric symbol in a single line
- `PUSHC name` acts like `PUSHC` followed by `SETCHARMAP name`
- `PUSHS ...` acts like `PUSHS` followed by `SECTION ...`
- `PUSHO ...` acts like `PUSHO` followed by `OPT ...`
- `cpl` can also be written as `cpl a`
- RGBLINK supports e.g. `ROMX FLOATING` in linker scripts, which allows assigning a section to no particular bank (this is generally only useful with SDCC object files)
- RGBLINK's `.sym` file now includes exported numeric constants (if this breaks your tools, please check [the `.sym` specification](https://rgbds.gbdev.io/sym/)!)
- RGBGFX `-i/--input-tileset` uses a given tileset instead of generating all the tiles itself
- RGBGFX `-r 0` infers a valid width for the reversed image
- RGBGFX `-X` or `-Y` only mirror tiles horizontally or vertically (`-X -Y` acts like `-m`)
- The [contrib/view_palettes.sh](https://github.com/gbdev/rgbds/blob/v0.9.0/contrib/view_palettes.sh) script lets you render a GBC `.pal` file as an image

## Changed

- `LOW(Label)` can be constant if `Label` is aligned to 8 or more bits
- `!expr` can be constant 0 if `expr` has any non-zero bits
- Error messages for undefined symbols mention if the symbol was previously purged
- Sections that grew too big are reported after the whole assembly file has been processed
- When RGBLINK encounters an unknown symbol, it lists the potential un-exported symbols by that name
- RGBLINK reports source locations for conflicts between object files
- RGBGFX `-r` gives a detailed warning if `-c` and `-p` mismatch; improved several other error messages too
- `LOW()` and `HIGH()` have their own RPN operators in the object file (this should lead to nicer output from [`rgbobj`](https://crates.io/crate/rgbobj))
- Clarified a lot of documentation, including for [fixed-point numbers](https://rgbds.gbdev.io/docs/v0.9.0/rgbasm.5#Fixed-point_expressions) and [macros](https://rgbds.gbdev.io/docs/v0.9.0/rgbasm.5#THE_MACRO_LANGUAGE)
- Explicitly document that `FMOD` behaves differently than `%`
- Increment object file revision number to 11

## Fixed

- `Label & CONSTANT` was not actually doing the `&` operation!
- Some nonsensical uses of empty local labels (`Global.`) were ignored
- Many bugs prevented SDCC object files from being processed correctly (special thanks to @RubenZwietering for helping with bugfixing and testing!)
- Crash when assembling a nonexistent file
- Crash when `INCLUDE`ing an unbounded-size file (e.g. `/dev/urandom`)
- Crash when passing an object file with invalid patch offsets to RGBLINK
- Crash when specifying palette files within RGBGFX at-files
- Crash when reversing an image if the tilemap referred to nonexistent tiles
- Multiple bugs with parsing text palette files (GPL, HEX, and PSP)
- Palette generation was suboptimal, potentially failing to fit within 8 palettes
- Maximum number of palettes wrapped around after 255
- Allow text palette files to have the maximum number of colors
- Building works on systems with bison 3.8.0

<hr /><em>This discussion was created from the release <a href='https://github.com/gbdev/rgbds/releases/tag/v0.9.0-rc1'>v0.9.0-rc1</a>.</em>

**Comment by Rangi42** [comment]
0.9.0-rc2 has been out for a while, and the final release is expected this month.
